### PR TITLE
Proxy: copy bidirectional fork

### PR DESCRIPTION
**Pull request by khanova**
## Problem

`tokio::io::copy_bidirectional` doesn't close the connection once one of the sides closes it. It's not really suitable for the postgres protocol.

## Summary of changes

Fork `copy_bidirectional` and initiate a shutdown for both connections.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [x] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by stradig** [comment]
Hi Anna! I think this should be called "copy_bidirectional"

**Comment by khanova** [comment]
> Hi Anna! I think this should be called "copy_bidirectional"

Oops, thanks, fixed!

File: proxy/src/proxy/copy_biderectional.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+#[derive(Debug)]
+pub(super) struct CopyBuffer {
+    read_done: bool,
+    need_flush: bool,
+    pos: usize,
+    cap: usize,
+    amt: u64,
+    buf: Box<[u8]>,
+}
+const DEFAULT_BUF_SIZE: usize = 8;
```
Comment (conradludgate) [inline proxy/src/proxy/copy_biderectional.rs:91]:
```suggestion
const DEFAULT_BUF_SIZE: usize = 8192;
```

Comment (khanova) [reply proxy/src/proxy/copy_biderectional.rs]: Oh, that was a testing artifact.

**Review by conradludgate — APPROVED** [review]

**Pull request merged**